File: 2019/day01.py
with open("../inputs/2019/day01.input", "r") as f:
    data = list(map(lambda x: int(x.strip()), f.readlines()))


def fuel(x):
    return max((x // 3) - 2, 0)


def total(x):
    req = fuel(x)
    total = 0
    while req > 0:
        total += req
        req = fuel(req)
    return total


def part1(data):
    return sum(map(fuel, data))


def part2(data):
    return sum(map(total, data))


print(f"Part 1: {part1(data)}")
print(f"Part 2: {part2(data)}")

File: rust-2022/src/day01.rs
pub fn read(inp: &[u8]) -> Vec<u32> {
    let mut v = Vec::with_capacity(256);
    let mut val = 0;
    for line in inp.split(|&c| c == b'\n') {
        if line.is_empty() {
            v.push(val);
            val = 0;
        } else {
            val += line
                .iter()
                .fold(0u32, |acc, &c| acc * 10 + ((c - b'0') as u32));
        }
    }
    v
}

pub fn part1(inp: &[u32]) -> u32 {
    inp.iter().fold(u32::MIN, |acc, &i| acc.max(i))
}

#[inline]
fn sort(r: &mut [u32]) {
    if r[0] > r[1] {
        r.swap(0, 1)
    }
    if r[0] > r[2] {
        r.swap(0, 2)
    }
    if r[1] > r[2] {
        r.swap(1, 2)
    }
}

pub fn part2(inp: &[u32]) -> u32 {
    const N: usize = 3;
    let mut r = [0; N];
    for &e in inp.iter() {
        if r[0] < e {
            r[0] = e;
            sort(&mut r);
        }
    }
    r.into_iter().sum()
}

pub fn run() -> (String, String) {
    let inp = include_bytes!(concat!(env!("CARGO_MANIFEST_DIR"), "/inputs/day01.input"));
    let data = read(inp);
    let a = part1(&data).to_string();
    let b = part2(&data).to_string();
    (a, b)
}
